pkg/kubelet/cm/memorymanager/memory_manager.go

```diff
@@ -0,0 +1,408 @@
+	var policy Policy
+
+	switch policyType(policyName) {
+
+	case policyTypeNone:
+		policy = NewPolicyNone()
+
+	case policyTypeStatic:
+		systemReserved, err := getSystemReservedMemory(machineInfo, nodeAllocatableReservation, reservedMemory)
+		if err != nil {
+			return nil, err
+		}
+
+		policy, err = NewPolicyStatic(machineInfo, systemReserved, affinity)
+		if err != nil {
+			return nil, err
+		}
+
+	default:
+		return nil, fmt.Errorf("unknown policy: \"%s\"", policyName)
+	}
+
+	manager := &manager{
+		policy:                     policy,
+		nodeAllocatableReservation: nodeAllocatableReservation,
+		stateFileDirectory:         stateFileDirectory,
+	}
+	manager.sourcesReady = &sourcesReadyStub{}
+	return manager, nil
+}
+
+// Start starts the memory manager reconcile loop under the kubelet to keep state updated
+func (m *manager) Start(activePods ActivePodsFunc, sourcesReady config.SourcesReady, podStatusProvider status.PodStatusProvider, containerRuntime runtimeService, initialContainers containermap.ContainerMap) error {
+	klog.Infof("[memorymanager] starting with %s policy", m.policy.Name())
+	m.sourcesReady = sourcesReady
+	m.activePods = activePods
+	m.podStatusProvider = podStatusProvider
+	m.containerRuntime = containerRuntime
+	m.containerMap = initialContainers
+
+	stateImpl, err := state.NewCheckpointState(m.stateFileDirectory, memoryManagerStateFileName, m.policy.Name(), m.containerMap)
+	if err != nil {
+		klog.Errorf("[memorymanager] could not initialize checkpoint manager: %v, please drain node and remove policy state file", err)
+		return err
+	}
+	m.state = stateImpl
+
+	err = m.policy.Start(m.state)
+	if err != nil {
+		klog.Errorf("[memorymanager] policy start error: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// AddContainer saves the value of requested memory for the guaranteed pod under the state and set memory affinity according to the topolgy manager
+func (m *manager) AddContainer(pod *v1.Pod, container *v1.Container, containerID string) error {
+	m.Lock()
+	m.containerMap.Add(string(pod.UID), container.Name, containerID)
+	m.Unlock()
+
+	// Get NUMA node affinity of blocks assigned to the container during Allocate()
+	var nodes []string
+	for _, block := range m.state.GetMemoryBlocks(string(pod.UID), container.Name) {
+		for _, nodeID := range block.NUMAAffinity {
+			nodes = append(nodes, strconv.Itoa(nodeID))
+		}
+	}
+
+	if len(nodes) < 1 {
+		klog.V(5).Infof("[memorymanager] update container resources is skipped due to memory blocks are empty")
+		return nil
+	}
+
+	affinity := strings.Join(nodes, ",")
+	klog.Infof("[memorymanager] Set container %q cpuset.mems to %q", containerID, affinity)
+	err := m.containerRuntime.UpdateContainerResources(containerID, &runtimeapi.LinuxContainerResources{CpusetMems: affinity})
```

Comment:
i dont have a great solution here in my head, but i find it confusing that we use a patch like syntax for updating container resources across cpu and memory managers.

Reply:
Can you please clarify, why does it confusing?

Reply:
ideally, we would have a way to roll-up our desired state and make a single call, but like i said earlier, its not clear how to do that in the current factoring.  not blocking on that, just more a comment.

Reply:
Thanks for the clarification, as part of our future work I will take a look at the whole flow again, and will try to figure out how can we to improve it.

Reply:
With the introduction of the `Allocate()` abstraction in `v1.18` it would actually be possible to do this without the `Update()` call now.

in the past we didn't know which resources we were going to give the container until we were inside the body of this `AddContainer()` call, so we had to use `Update()` since the container was already started.

With the introduction of the `Allocate()` abstraction, we now know which resources will be allocated to each container at pod admission time before the container ever starts.

We would need to add some plumbing to propagate this information up to the initial `ContainerCreate()` call, but the information is already available at the right time.